File: cpp/3558.number-of-ways-to-assign-edge-weights-i.cpp
#include <vector>
#include <unordered_map>

using std::vector, std::unordered_map;

class Solution {
public:
    int assignEdgeWeights(vector<vector<int>>& edges) {
      const int MOD = int(1e9) + 7;

      const int n = edges.size() + 2;
      // 3 edges: 2 childs + 1 parent
      vector<vector<int>> adj(n); 

      for (vector<int>& edge : edges ) {
        adj[edge[0]].push_back(edge[1]);
        adj[edge[1]].push_back(edge[0]);
      }

      auto dfs = [&adj](this auto&& self, int node, int parent) -> int {
        int d = 0;

        for (int child : adj[node]) {
          if (child == parent || child == 0) {
            continue;
          }

          d = std::max(d, 1 + self(child, node));
        }

        return d;
      };

      auto mod_pow = [](long long n, long long m) -> long long {
        long long result = 1;
        long long base = 2 % m;

        while (n > 0) {
          if (n & 1) {
            result = (result * base) % m;
          }

          base = (base * base) % m;
          n >>= 1;
        }

        return result;
      };

      int k = dfs(1, -1);

      return mod_pow(k - 1, MOD);
    }
};

File: python/3558.number-of-ways-to-assign-edge-weights-i.py
class Solution:
    def assignEdgeWeights(self, edges: list[list[int]]) -> int:
        # number of edges in tree is equal number of nodes - 1
        n = len(edges) + 1 # number of nodes
        MOD = int(1e9) + 7

        adj = [[] for _ in range(n)]
        for a, b, *_ in edges:
            # sort and shift from 1-base indexes to 0-base
            u, v = min(a, b) - 1, max(a, b) - 1
            adj[u].append(v) 

        def bfs_max_depth(start: int) -> int:
            q = deque([start])
            d = [-1] * n
            d[start] = 0
            while q:
                node = q.popleft()
                for child in adj[node]:
                    if d[child] != -1:
                        continue
                    
                    d[child] = d[node] + 1
                    q.append(child)
            
            return max(d)

        def dfs_max_depth(node: int, parent: int = -1) -> int:
            depth = 0

            for child in adj[node]:
                if child == parent:
                    continue
                
                depth = max(depth, 1 + dfs_max_depth(child, node))

            return depth
        
        k = bfs_max_depth(0)
        return pow(2, k - 1, MOD)